envisage/plugins/ipython_kernel/internal_ipkernel.py

```diff
@@ -90,62 +117,51 @@ def init_ipkernel(self, gui_backend):
         # Start IPython kernel with GUI event loop support
         self.ipkernel = gui_kernel(gui_backend)
 
-        # Since ipykernel 4.7, the io_loop attribute of the kernel is not
-        # initialized anymore
-        # Reference: https://github.com/enthought/envisage/issues/107
-        # Workaround: Retrieve the kernel on the IPykernelApp and set the
-        # io_loop without starting it!
-        if NEEDS_IOLOOP_PATCH and not hasattr(self.ipkernel.kernel, 'io_loop'):
-            self.ipkernel.kernel.io_loop = ioloop.IOLoop.instance()
```

Comment:
Can you comment on why this piece of code got removed? Was this fixed upstream?

Reply:
It was the wrong fix. The original problem that this code was fixing was triggered by this line: 

https://github.com/enthought/envisage/blob/0950c2dfef6a5b44c24fff1b20d71b087198fb97/envisage/plugins/ipython_kernel/internal_ipkernel.py#L127

But that line should not have been there in the first place: the `exit_now` attribute is there to tell the shell to shut down under a running event loop, but by the time we reach this code the event loop has already been shut down. So the right fix was to remove this line (and make sure that any other necessary shell cleanup has taken place, which is what some of the rest of this PR does).